src/mlir/generator.h

```diff
@@ -40,58 +37,74 @@ namespace mlir::verona
    */
   class ReturnValue
   {
-    using ValuesTy = llvm::SmallVector<mlir::Value, 4>;
+    using ResultTy = std::variant<Value, ModuleOp, FuncOp>;
+    using ValuesTy = llvm::SmallVector<ResultTy, 1>;
     ValuesTy values;
 
   public:
     /// Default constructor, builds an empty return value.
     ReturnValue() {}
     /// Single value constructor.
-    ReturnValue(mlir::Value value)
+    ReturnValue(Value value)
     {
       values.push_back(value);
     }
+    /// Single value constructor.
+    ReturnValue(ModuleOp mod)
+    {
+      values.push_back(mod);
+    }
+    /// Single value constructor.
+    ReturnValue(FuncOp func)
+    {
+      values.push_back(func);
+    }
     /// Multiple value constructor.
-    ReturnValue(mlir::ResultRange& range)
+    ReturnValue(ResultRange& range)
     {
       values.insert(values.begin(), range.begin(), range.end());
     }
-    /// Assignment operator for mlir::Value.
-    ReturnValue& operator=(mlir::Value& value)
+    /// Assignment operator for Value.
+    ReturnValue& operator=(Value& value)
     {
       values.push_back(value);
       return *this;
     }
 
-    /// Returns true if this return value has exactly one value.
-    bool hasValue() const
+    /// Returns true if this return value has exactly `n` values.
+    bool hasValue(size_t n = 1) const
     {
-      return values.size() == 1;
+      assert(n != 0);
+      return values.size() == n;
     }
 
     /// Access to the single value held, error if none or more than one.
-    mlir::Value get() const
+    template<class T>
+    T get() const
     {
       assert(values.size() > 0 && "Access to empty return value");
       assert(values.size() == 1 && "Direct access to multiple values");
-      return values[0];
+      return std::get<T>(values[0]);
     }
 
-    /// Returns true if this return value has one or more values.
-    bool hasValues() const
+    /// Access a specific value held, error if none or not enough.
```

Comment:
I'll change the comment to "asserts". I don't want to create a full error handling now because I want it to be part of a diagnostics infrastructure and not make the same mistake we did last time, to bake in the error handling too soon.